dask/array/tests/test_creation.py

```diff
@@ -407,32 +407,34 @@ def test_eye():
         assert 4 < x.npartitions < 32
 
 
-def test_diag():
+@pytest.mark.parametrize("k", [0, 3, -3, 8])
+def test_diag(k):
     v = np.arange(11)
-    assert_eq(da.diag(v), np.diag(v))
+    assert_eq(da.diag(v, k), np.diag(v, k))
 
     v = da.arange(11, chunks=3)
-    darr = da.diag(v)
-    nparr = np.diag(v)
+    darr = da.diag(v, k)
+    nparr = np.diag(v, k)
     assert_eq(darr, nparr)
-    assert sorted(da.diag(v).dask) == sorted(da.diag(v).dask)
+    assert sorted(da.diag(v, k).dask) == sorted(da.diag(v, k).dask)
 
     v = v + v + 3
-    darr = da.diag(v)
-    nparr = np.diag(v)
+    darr = da.diag(v, k)
+    nparr = np.diag(v, k)
     assert_eq(darr, nparr)
 
     v = da.arange(11, chunks=11)
-    darr = da.diag(v)
-    nparr = np.diag(v)
+    darr = da.diag(v, k)
+    nparr = np.diag(v, k)
     assert_eq(darr, nparr)
-    assert sorted(da.diag(v).dask) == sorted(da.diag(v).dask)
+    assert sorted(da.diag(v, k).dask) == sorted(da.diag(v, k).dask)
 
     x = np.arange(64).reshape((8, 8))
-    assert_eq(da.diag(x), np.diag(x))
-
+    assert_eq(da.diag(x, k), np.diag(x, k))
     d = da.from_array(x, chunks=(4, 4))
-    assert_eq(da.diag(d), np.diag(x))
+    assert_eq(da.diag(d, k), np.diag(x, k))
+    d = da.from_array(x, chunks=((3, 2, 3), (4, 4)))
+    assert_eq(da.diag(d, k), np.diag(x, k))
```

Comment:
Can you add a test with a rectangular input array as well? I only see square input arrays above.

While you're at it, would you mind splitting up the tests here a bit to separate the two uses of `diag`? Putting them together in the same test makes it a bit hard to read.

- One (or more, up to you) tests for creating a diagonal array from a 1d array
- One (or more, up to you) tests for extracting the diagonal from a 2d array

Reply:
Good point.  I'll do that.

Reply:
I've added and reorganized the tests plus ensured full coverage of `diag()` & `diagonal()`

Reply:
@jcrist 

CI tests failures seem to reveal that `cupy` type input arrays to `diagonal()` need special handling.  I'm taking a closer look at this.

Reply:
@jcrist

CI tests for `cupy`-type arrays have now passed!